### PR TITLE
Fix binary size.

### LINKED ISSUE #377: Binary size increase between v9.1.0 and v9.2.0
Our latest minor release has increased in binary size due to a missing LTO configuration. We are working on resolving this and will publish a v9.2.1 patch asap.

cc @mapbox/maps-android

### PR DESCRIPTION
**Pull request by tarigo**
Add compiler and linker options to reduce binary size for release build:
  - reduce symbol visibility: `-fvisibility=hidden`, `-fvisibility-inlines=hidden`;
  - link time optimization `-flto` and enable gold plugin;
  - remove duplicated code: `–icf=all`;
  - size optimization for compilation (`-Oz`);
  - garbage collect unused sections: `-Wl,--gc-sections`

Fixes #377.

### CODE REVIEW

File: MapboxGLAndroidSDK/src/cpp/CMakeLists.txt [hunk 1]
```diff
@@ -1,208 +1,212 @@
...
-  mapbox-gl
-  PRIVATE
-        mbgl-compiler-options
-        Mapbox::Base::jni.hpp
-        mbgl-core
-        mbgl-vendor-unique_resource
-
-)
+target_link_libraries(mapbox-gl
+                      PRIVATE $<$<CONFIG:Release>:-O2>
```
Comment (pozdnyakov) [inline MapboxGLAndroidSDK/src/cpp/CMakeLists.txt:203]: -Os ?

**Comment by alexshalamov** [comment]
@tarigo what is the diff between 9.2.0 and this PR?

File: MapboxGLAndroidSDK/src/cpp/CMakeLists.txt (hunk 1 above)
Comment (tarigo) [reply MapboxGLAndroidSDK/src/cpp/CMakeLists.txt]: Right.

Comment (alexshalamov) [reply MapboxGLAndroidSDK/src/cpp/CMakeLists.txt]: Was it fixed in NDK? There was a long standing bug that -Os and -Oz did not work, 🤞 if -Oz works.

Comment (tarigo) [reply MapboxGLAndroidSDK/src/cpp/CMakeLists.txt]: Well, for ` mbgl-core`  -Oz for object files and -O2 for linker works. Let's see -Oz combination for both. :-)

**Comment by tarigo** [comment]
> @tarigo what is the diff between 9.2.0 and this PR?

`aar`: 8,8M/8,2M  -7.4%
`libmapbox-gl.so` arm64:  6,1M/4,5M (-33.6%) 
`libmapbox-gl.so` arm:   4,7M/3,1M (-33.6%)

File: MapboxGLAndroidSDK/src/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -1,208 +1,212 @@
...
-  mapbox-gl
-  PRIVATE
-        mbgl-compiler-options
-        Mapbox::Base::jni.hpp
-        mbgl-core
-        mbgl-vendor-unique_resource
-
-)
+target_link_libraries(mapbox-gl
+                      PRIVATE $<$<CONFIG:Release>:-Os>
```
Comment (pozdnyakov) [inline MapboxGLAndroidSDK/src/cpp/CMakeLists.txt:203]: Oz ? :smile:

Comment (tarigo) [reply MapboxGLAndroidSDK/src/cpp/CMakeLists.txt]: Ops, typo, fixed!

**Pull request merged**